### PR TITLE
Don't reset site.url to localhost:4000 by default

### PR DESCRIPTION
**Pull request by ashmaroli**
Local tests show that the local preview server is based on `config["host"]` and `config["port"]` values..
So even if `config["url"] == https://jekyllrb.com`, the site will be served at the default setting: `//127.0.0.1:4000`

This simplifies users' workflow by **not having to `build` the site again with `JEKYLL_ENV=production`**
(unless their templates are designed to render differently for the two ENV vars, which is a different story altogether)

*P.S. The above observation was made on Windows. Please confirm for Unix platforms, if possible.*

### CODE REVIEW

File: lib/jekyll/commands/serve.rb [hunk 1]
```diff
@@ -249,9 +249,12 @@ def format_url(ssl_enabled, address, port, baseurl = nil)
 
         def default_url(opts)
           config = configuration_from_options(opts)
+          auth = config.values_at("host", "port").join(":")
+          return config["url"] if auth == "127.0.0.1:4000"
+
           format_url(
             config["ssl_cert"] && config["ssl_key"],
-            config["host"] == "127.0.0.1" ? "localhost" : config["host"],
```
Comment (pathawks) [inline lib/jekyll/commands/serve.rb]: `localhost` is nicer than an IP address

**Comment by ashmaroli** [comment]
@DirtyF  Did you give this branch a test-run locally..?

**Comment by DirtyF** [comment]
@ashmaroli I tested jekyll's website locally and `url` equals `site.url` in `feed.xml` and `robots.txt` without having to specify production environment on the command line.

**Comment by DirtyF** [comment]
@jekyll: merge +minor

**Comment by DirtyF** [comment]
@ashmaroli serve with livereload is now failing on `master`:

```
 Auto-regeneration: enabled for '/Users/frank/code/jekyll/jekyll/docs'
Configuration file: /Users/frank/code/jekyll/jekyll/docs/_config.yml
rake aborted!
Errno::EADDRINUSE: Address already in use - bind(2) for 127.0.0.1:4000
/Users/frank/code/jekyll/jekyll/lib/jekyll/commands/serve.rb:215:in `new'
/Users/frank/code/jekyll/jekyll/lib/jekyll/commands/serve.rb:215:in `start_up_webrick'
/Users/frank/code/jekyll/jekyll/lib/jekyll/commands/serve.rb:102:in `process'
/Users/frank/code/jekyll/jekyll/rake/site.rake:36:in `block (2 levels) in <top (required)>'
/Users/frank/code/jekyll/jekyll/vendor/bundle/ruby/2.6.0/gems/rake-13.0.1/exe/rake:27:in `<top (required)>'
/Users/frank/.gem/ruby/2.6.0/gems/bundler-2.2.0.rc.2/lib/bundler/cli/exec.rb:63:in `load'
/Users/frank/.gem/ruby/2.6.0/gems/bundler-2.2.0.rc.2/lib/bundler/cli/exec.rb:63:in `kernel_load'
/Users/frank/.gem/ruby/2.6.0/gems/bundler-2.2.0.rc.2/lib/bundler/cli/exec.rb:28:in `run'
/Users/frank/.gem/ruby/2.6.0/gems/bundler-2.2.0.rc.2/lib/bundler/cli.rb:494:in `exec'
/Users/frank/.gem/ruby/2.6.0/gems/bundler-2.2.0.rc.2/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/Users/frank/.gem/ruby/2.6.0/gems/bundler-2.2.0.rc.2/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
/Users/frank/.gem/ruby/2.6.0/gems/bundler-2.2.0.rc.2/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
/Users/frank/.gem/ruby/2.6.0/gems/bundler-2.2.0.rc.2/lib/bundler/cli.rb:30:in `dispatch'
/Users/frank/.gem/ruby/2.6.0/gems/bundler-2.2.0.rc.2/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
/Users/frank/.gem/ruby/2.6.0/gems/bundler-2.2.0.rc.2/lib/bundler/cli.rb:24:in `start'
/Users/frank/.gem/ruby/2.6.0/gems/bundler-2.2.0.rc.2/exe/bundle:49:in `block in <top (required)>'
/Users/frank/.gem/ruby/2.6.0/gems/bundler-2.2.0.rc.2/lib/bundler/friendly_errors.rb:117:in `with_friendly_errors'
/Users/frank/.gem/ruby/2.6.0/gems/bundler-2.2.0.rc.2/exe/bundle:37:in `<top (required)>'
/Users/frank/.gem/ruby/2.6.0/bin/bundle:23:in `load'
/Users/frank/.gem/ruby/2.6.0/bin/bundle:23:in `<main>'
Tasks: TOP => site:preview
```

**Comment by ashmaroli** [comment]
Is another instance already being served?
```
Errno::EADDRINUSE: Address already in use - bind(2) for 127.0.0.1:4000
```

**Comment by DirtyF** [comment]
No, and I get the same error when running without livereload:

```
bundler exec jekyll serve --trace
bundler: failed to load command: jekyll (/Users/frank/code/jekyll/jekyll/vendor/bundle/ruby/2.6.0/bin/jekyll)
Errno::EADDRINUSE: Address already in use - bind(2) for 127.0.0.1:4000
  /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/socket.rb:201:in `bind'
  /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/socket.rb:201:in `listen'
  /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/socket.rb:764:in `block in tcp_server_sockets'
  /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/socket.rb:227:in `each'
  /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/socket.rb:227:in `foreach'
  /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/socket.rb:762:in `tcp_server_sockets'
  /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/webrick/utils.rb:65:in `create_listeners'
  /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/webrick/server.rb:127:in `listen'
  /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/webrick/server.rb:108:in `initialize'
  /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/webrick/httpserver.rb:47:in `initialize'
  /Users/frank/code/jekyll/jekyll/lib/jekyll/commands/serve.rb:215:in `new'
  /Users/frank/code/jekyll/jekyll/lib/jekyll/commands/serve.rb:215:in `start_up_webrick'
  /Users/frank/code/jekyll/jekyll/lib/jekyll/commands/serve.rb:102:in `process'
  /Users/frank/code/jekyll/jekyll/lib/jekyll/command.rb:91:in `block in process_with_graceful_fail'
  /Users/frank/code/jekyll/jekyll/lib/jekyll/command.rb:91:in `each'
  /Users/frank/code/jekyll/jekyll/lib/jekyll/command.rb:91:in `process_with_graceful_fail'
  /Users/frank/code/jekyll/jekyll/lib/jekyll/commands/serve.rb:86:in `block (2 levels) in init_with_program'
  /Users/frank/code/jekyll/jekyll/vendor/bundle/ruby/2.6.0/gems/mercenary-0.4.0/lib/mercenary/command.rb:221:in `block in execute'
  /Users/frank/code/jekyll/jekyll/vendor/bundle/ruby/2.6.0/gems/mercenary-0.4.0/lib/mercenary/command.rb:221:in `each'
  /Users/frank/code/jekyll/jekyll/vendor/bundle/ruby/2.6.0/gems/mercenary-0.4.0/lib/mercenary/command.rb:221:in `execute'
  /Users/frank/code/jekyll/jekyll/vendor/bundle/ruby/2.6.0/gems/mercenary-0.4.0/lib/mercenary/program.rb:44:in `go'
  /Users/frank/code/jekyll/jekyll/vendor/bundle/ruby/2.6.0/gems/mercenary-0.4.0/lib/mercenary.rb:21:in `program'
  /Users/frank/code/jekyll/jekyll/exe/jekyll:15:in `<top (required)>'
  /Users/frank/code/jekyll/jekyll/vendor/bundle/ruby/2.6.0/bin/jekyll:23:in `load'
  /Users/frank/code/jekyll/jekyll/vendor/bundle/ruby/2.6.0/bin/jekyll:23:in `<top (required)>'
```

**Comment by benlk** [comment]
If you, like me, have found this ticket while trying to debug `absolute_url` using the wrong URL under `bundle exec jekyll serve`, there's a command-line workaround documented in https://github.com/jekyll/jekyll/issues/8541

**Comment by DirtyF** [comment]
@ashmaroli why don't we only do that when passing ENV=production?

If ENV=production jekyll serve + absolute URL == your site.url
else jekyll serve defaults to localhost

**Comment by benlk** [comment]
If conversation is going to be reopened on whether or not this change was the right move, I'll add my two cents:

> This simplifies users' workflow by not having to build the site again with JEKYLL_ENV=production
(unless their templates are designed to render differently for the two ENV vars, which is a different story altogether)

When running `bundle exec jekyll serve`, with asset URLs piped through the `absolute_url` filter, under Jekyll 4.2, my localhost development site served at `http://localhost:4000/` attempts to load assets like images, stylesheets, scripts, fonts and so on via the production URL `https://example.com/`. If those assets exist on the production server, then they load.

However, this PR's change leads to problems in the following cases:
- images not yet uploaded to the production server do not appear, instead failing with `404`s
- stylesheets on prod don't have changes that were made to the `http://localhost:4000/` version of the file
- scripts on prod don't match markup served from `http://localhost:4000/`
- font files fail to load because the production server doesn't supply a CORS header that allows their use on `http://localhost:4000/`

To avoid the problem of deploying files with `http://localhost:4000/` URLs to `https://example.com/`, my deploy script runs `jekyll clean` before `jekyll build`, and then uploads to the production server.

I often run `bundle exec jekyll serve` with `--drafts --unpublished`. Even if I changed all my asset URL handling to be relative URLs, I would still `jekyll clean` to avoid uploading draft or unpublished work to the production site.

This change doesn't help me. All it does is change my default local development command from `bundle exec jekyll serve` to `bundle exec jekyll serve --host localhost --port 4000` — it's not a terrible bother, but it took me three hours to find what had changed.

----

Whether or not this change is undone, https://jekyllrb.com/docs/configuration/options/#serve-command-options should be updated to:
- describe the effects of `serve` upon the variables it treats differently
- state that `serve` serves from the same build directory used by `jekyll build`, and encourage users to run `jekyll build` before deploying files from that directory to a production server.

**Comment by evanwill** [comment]
I also just don't quite understand the justification for the change. 
I thought the purpose of using `absolute_url` or `site.url` in templates *was* that it would be automatically swapped out on `jekyll s` vs. `jekyll build`? 
The change definitely breaks the majority of my older projects, so I will have to hold them all back at Jekyll 4.1.1 or explain to users that you need to add `--port 4001` to revert to expected behavior. 

As @benlk mentions, in either case it would be helpful to clarify the documentation.

**Comment by ChaelCodes** [comment]
Documentation should be updated when major changes like this occur. The [variables](https://jekyllrb.com/docs/variables/) documentation makes some promises about the behavior of site.url that are no longer true.

**Comment by benlk** [comment]
I have filed a PR to revert these changes: https://github.com/jekyll/jekyll/pull/8620

Sorry for the delay in getting them to you.

**Pull request merged**